protocols/bftcosi/bftcosi.go

```diff
@@ -187,291 +200,300 @@ func (bft *ProtocolBFTCoSi) Dispatch() error {
+// the threshold of participants refused to sign
 func (bft *ProtocolBFTCoSi) handleChallengeCommit(ch *ChallengeCommit) error {
-	ch.Challenge = bft.commit.Challenge(ch.Challenge)
-	hash := bft.Suite().Hash()
-	hash.Write(bft.Msg)
-	h := hash.Sum(nil)
+	if !bft.IsRoot() {
+		bft.commit.Challenge(ch.Challenge)
+	}
 
 	// verify if the signature is correct
-	if err := cosi.VerifyCosiSignatureWithException(bft.suite,
-		bft.AggregatedPublic, h, ch.Signature,
-		ch.Exceptions); err != nil {
-		dbg.Error(bft.Name(), "Verification of the signature failed:", err)
+	if err := VerifyBFTSignature(bft.suite, ch.Signature, bft.Roster().Publics()); err != nil {
+		dbg.Lvl2(bft.Name(), "Verification of the signature failed:", err)
 		bft.signRefusal = true
 	}
 
-	// Check if we have no more than 1/3 failed nodes
-	if len(ch.Exceptions) > int(bft.threshold) {
-		dbg.Errorf("More than 1/3 (%d/%d) refused to sign ! ABORT",
-			len(ch.Exceptions), len(bft.Roster().List))
+	// Check if we have no more than threshold failed nodes
+	if len(ch.Signature.Exceptions) >= int(bft.threshold) {
+		dbg.Lvlf2("%s: More than threshold (%d/%d) refused to sign - aborting.",
+			bft.Roster(), len(ch.Signature.Exceptions), len(bft.Roster().List))
 		bft.signRefusal = true
 	}
 
 	// store the exceptions for later usage
-	bft.tempExceptions = ch.Exceptions
+	bft.tempExceptions = ch.Signature.Exceptions
 	dbg.Lvl4("BFTCoSi handle Challenge COMMIT")
+
 	if bft.IsLeaf() {
-		return bft.startResponseCommit()
+		return bft.handleResponseCommit(nil)
 	}
 
-	if err := bft.SendToChildrenInParallel(ch); err != nil {
-		dbg.Error(err)
+	return bft.SendToChildrenInParallel(ch)
+}
+
+// startResponse dispatches the response to the correct round-type
+func (bft *ProtocolBFTCoSi) startResponse(t RoundType, r *Response) error {
+	if t == RoundPrepare {
+		return bft.handleResponsePrepare(r)
 	}
-	return nil
+	return bft.handleResponseCommit(r)
 }
 
-// startPrepareResponse wait the verification of the block and then start the
-// challenge process
-func (bft *ProtocolBFTCoSi) startResponsePrepare() error {
-	// create response
-	resp, err := bft.prepare.CreateResponse()
+// challenge process.
+// If 'r' is nil, it will starts the response process.
+func (bft *ProtocolBFTCoSi) handleResponsePrepare(r *Response) error {
+	if r != nil {
+		// check if we have enough responses
+		bft.tprMut.Lock()
+		defer bft.tprMut.Unlock()
+		bft.tempPrepareResponse = append(bft.tempPrepareResponse, r.Response)
+		bft.tempExceptions = append(bft.tempExceptions, r.Exceptions...)
+		if len(bft.tempPrepareResponse) < len(bft.Children()) {
+			return nil
+		}
+	}
+
+	// wait for verification
+	bzrReturn, ok := bft.waitResponseVerification()
+	// append response
+	if !ok {
+		dbg.Lvl3(bft.Roster(), "Refused to sign")
+	}
+
+	// Return if we're not root
+	if !bft.IsRoot() {
+		return bft.SendTo(bft.Parent(), bzrReturn)
+	}
+	// let's modify the cosi signature so it don't account the exception
```

Comment:
using `cosi.Signature()` returns the full signature with *all* responses aggregated inside. Because cosi does not support exceptions for the moment, we have to remove the *responses* from the signature which should not be there (i.e. exceptions).